File: sdk/lib/lilka/src/lilka/alert.cpp
#include "ui.h"

namespace lilka {

Alert::Alert(String title, String message) {
    this->title = title;
    this->message = message;
    this->done = false;
    this->button = Button::COUNT;
    activationButtons.push_back(Button::A);
}

void Alert::setTitle(String title) {
    this->title = title;
}

void Alert::setMessage(String message) {
    this->message = message;
}

void Alert::update() {
    State state = controller.getState();
    lilka::_StateButtons& buttonsArray = *reinterpret_cast<lilka::_StateButtons*>(&state);
    std::vector<Button>::iterator it =
        std::find_if(activationButtons.begin(), activationButtons.end(), [&buttonsArray](Button activationButton) {
            return buttonsArray[activationButton].justPressed;
        });
    if (it != activationButtons.end()) {
        button = *it;
        done = true;
    }
    vTaskDelay(LILKA_UI_UPDATE_DELAY_MS / portTICK_PERIOD_MS);
}

void Alert::draw(Arduino_GFX* canvas) {
    int top = canvas->height() / 8;
    int mid = canvas->height() / 8 * 2;
    int bottom = canvas->height() / 8 * 7;

    int left = canvas->width() / 8;
    int right = canvas->width() / 8 * 7;
    int width = right - left;
    int xMargin = 4;

    canvas->setTextColor(lilka::colors::White);

    canvas->fillRect(left, top, width, mid - top, lilka::colors::Midnight_blue);
    canvas->setFont(FONT_6x13);
    canvas->setTextSize(2);
    canvas->setTextBound(left + xMargin, top, width - xMargin * 2, mid - top);
    canvas->setCursor(left + xMargin, top + 13 * 2);
    canvas->println(title);

    canvas->fillRect(left, mid, width, bottom - mid, canvas->color565(32, 96, 96));
    canvas->setFont(FONT_9x15);
    canvas->setTextSize(1);
    canvas->setTextBound(left + xMargin, mid, width - xMargin * 2, bottom - mid);
    canvas->setCursor(left + xMargin, mid + 20);
    canvas->println(message);
}

bool Alert::isFinished() {
    if (done) {
        done = false;
        return true;
    }
    return false;
}

void Alert::addActivationButton(Button activationButton) {
    activationButtons.push_back(activationButton);
}

Button Alert::getButton() {
    return button;
}

} // namespace lilka

File: firmware/keira/sdcard/rf/libs/alert.lua
--[[
    Alert
    Автор: Андрій Ситников (drronnie85)
    Опис: Клас для малювання сповіщень
--]]

WHITE = display.color565(255, 255, 255)
MIDNIGHT_BLUE = display.color565(25, 25, 112)

Alert = {
    title = nil,
    message = nil
}

function Alert:new(title, message)
    obj = {
        title = title,
        message = message
    }
    setmetatable(obj, self)
    self.__index = self
    return obj
end

function Alert:draw(delta)
    local top = display.height / 8
    local mid = display.height / 8 * 2
    local bottom = display.height / 8 * 7

    local left = display.width / 8
    local right = display.width / 8 * 7
    local width = right - left
    local xMargin = 4

    title = self.title or "Інформація"
    message = self.message or ""

    display.set_text_color(WHITE)
    display.fill_rect(left, top, width, mid - top, MIDNIGHT_BLUE)
    display.set_font("6x13")
    display.set_text_size(2)
    display.set_text_bound(left + xMargin, top, width - xMargin * 2, mid - top)
    display.set_cursor(left + xMargin, top + 13 * 2)
    display.print(title)
    display.fill_rect(left, mid, width, bottom - mid, display.color565(32, 96, 96))
    display.set_font("9x15")
    display.set_text_size(1)
    display.set_text_bound(left + xMargin, mid, width - xMargin * 2, bottom - mid)
    display.set_cursor(left + xMargin, mid + 20)
    display.print(message);

end

return Alert